src/Symfony/Component/Messenger/Transport/Sender/SendersLocator.php

```diff
@@ -23,17 +27,29 @@
  */
 class SendersLocator implements SendersLocatorInterface
 {
-    private $senders;
+    private $sendersMap;
+    private $sendersLocator;
+    private $useLegacyLookup = false;
     private $sendAndHandle;
 
     /**
-     * @param SenderInterface[][] $senders
-     * @param bool[]              $sendAndHandle
+     * @param string[][] $sendersMap An array, keyed by "type", set to an array of sender aliases
+     * @param ContainerInterface Locator of senders, keyed by sender alias
+     * @param bool[]     $sendAndHandle
      */
-    public function __construct(array $senders, array $sendAndHandle = [])
+    public function __construct(array $sendersMap, /*ContainerInterface*/ $sendersLocator = null, array $sendAndHandle = [])
```

Comment:
This is really the "messiest" part of the PR. Because we need to send to the failure transport, identified by its "name", we need a way to ask the system: "please give me the sender named `my_failed_transport`". The new `getSenderByAlias()` accomplishes this. And it works great, but it was a "shift" in the system, as - until now - we only ever asked "Give me the senders for this envelope/message".